### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by Tyu-cr**

### CODE REVIEW

**Comment by evermake** [comment]
@Tyu-cr, how things are going? Do you need some help?

**Comment by Tyu-cr** [comment]
No. I will push all changes toworrow i guess. Almost done

**Comment by evermake** [comment]
image

@Tyu-cr , have you installed `jwt` package with `pip`? We use `poetry` as the project's dependency manager.

File: app/api/deps.py [hunk 1]
```diff
@@ -1,28 +1,100 @@
...
+        raise TypeError("token must be a string")
+
+    try:
+        access_token, refresh_token = token.split(" ")
+        if access_token is None:
+            raise CredentialsException(detail="No access token found")
+        if refresh_token is None:
+            raise CredentialsException(detail="No refresh token found")
+
+        if check_token_expired(access_token, "access"):
```
Comment (evermake) [inline app/api/deps.py:41]: Please, read PyJWT documentation carefully. Especially [this](https://pyjwt.readthedocs.io/en/latest/usage.html#expiration-time-claim-exp). Why reinvent the wheel?

File: app/api/iam/router.py [hunk 2]
```diff
@@ -1,3 +1,126 @@
...
+    user = authenticate_user(form_data.username, form_data.password)
+    if not user:
+        raise CredentialsException("Wrong email or password")
+
+    access_token = create_jwt_token(
+        email=form_data.username,
+        expires_delta=timedelta(minutes=ACCESS_TOKEN_LIFETIME_MINUTES),
+        token_type="access"
+    )
+    refresh_token = create_jwt_token(
```
Comment (evermake) [inline app/api/iam/router.py:47]: Refresh tokens should not be JWT tokens. It's just a secure randomly generated string, that should be stored in the system. Please, read the issue #8 , where I described everything and ask me, if something is unclear.

File: app/api/iam/schemas.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    access_token: str
+    token_type: str
+
+
+class User(BaseModel):
+    email_address: str | None = None
+    ...
+
+
+class UserPassword(User):
```
Comment (evermake) [inline app/api/iam/schemas.py:14]: 1. Why separate model?
2. Users will not have passwords 🙁

File: app/api/iam/utils.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")
+
+JWT_ALGORITHM = config.jwt_algorithm
+ACCESS_TOKEN_LIFETIME_MINUTES = config.access_token_lifetime_minutes
+REFRESH_TOKEN_LIFETIME_MINUTES = config.refresh_token_lifetime_minutes
+SECRET_ACCESS_KEY = config.secret_access_key
+SECRET_REFRESH_KEY = config.secret_refresh_key
+
+test_db = {
```
Comment (evermake) [inline app/api/iam/utils.py:20]: I think you wanted to implement an `AuthRepository` in `app.adapters`, but not put it there...

File: app/api/iam/utils.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        if token_type == "access":
+            payload = decode(token, SECRET_ACCESS_KEY, algorithms=JWT_ALGORITHM)
+        elif token_type == "refresh":
+            payload = decode(token, SECRET_REFRESH_KEY, algorithms=JWT_ALGORITHM)
+
+        expiration_time = payload.get("exp")
+        if expiration_time is None:
+            return True
+
+        # current_time = datetime.utcnow().timestamp()
```
Comment (evermake) [inline app/api/iam/utils.py:96]: Uncomment or remove code

**Review by evermake — CHANGES_REQUESTED** [review]

File: app/api/iam/router.py (hunk 2 above)
Comment (Tyu-cr) [reply app/api/iam/router.py:47]: What is the difference? jwt is basically the same thing, but with additional checks.

File: app/api/iam/utils.py (hunk 4 above)
Comment (Tyu-cr) [reply app/api/iam/utils.py:20]: Maybe, I'll ask you about it later.

File: app/api/iam/schemas.py (hunk 3 above)
Comment (Tyu-cr) [reply app/api/iam/schemas.py:14]: 2. How did they login tho, only email?🤨

File: app/config.py [hunk 6]
```diff
@@ -20,6 +22,16 @@ class Config(BaseSettings):
 
     cors_allowed_origins: list[AnyHttpUrl] = []
 
+    load_dotenv()
+    jwt_algorithm: str = os.getenv("JWT_ALGORITHM")
+    # openssl rand -hex 32
+    # TODO: encode keys in base64
```
[the remark this answers is not in the document]
Comment (Tyu-cr) [reply app/config.py:28]: Some stack overflow guys say it's even better. idk

File: app/api/iam/schemas.py (hunk 3 above)
Comment (evermake) [reply app/api/iam/schemas.py:14]: Read the issue #8 carefully again

File: app/api/iam/router.py (hunk 2 above)
Comment (evermake) [reply app/api/iam/router.py:47]: JWT tokens are meant to hold some information and not to be stored on backend.

We don't need to hold any information, we just need a long random string, which is stored on backend.

File: app/config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (evermake) [reply app/config.py]: It's okay to use one generated by `openssl rand -hex 32`

File: app/api/iam/router.py (hunk 2 above)
Comment (Tyu-cr) [reply app/api/iam/router.py:47]: Expiration time?

File: app/api/iam/schemas.py (hunk 3 above)
Comment (Tyu-cr) [reply app/api/iam/schemas.py:14]: oke

**Review by evermake — APPROVED** [review]

**Pull request merged**